benches/r1cs.rs

```diff
@@ -72,181 +69,113 @@ For K = 1:
 /// Enforces that the output variables `y` are a valid reordering of the inputs variables `x`.
 /// The inputs and outputs are all tuples of the `Variable, Assignment`, where the `Assignment`
 /// can be either assigned as `Value::Scalar` or unassigned as `Missing`.
-pub fn fill_cs<CS: ConstraintSystem>(
-    cs: &mut CS,
-    x: Vec<(Variable, Assignment)>,
-    y: Vec<(Variable, Assignment)>,
-) -> Result<(), KShuffleError> {
-    let one = Scalar::one();
-    let z = cs.challenge_scalar(b"k-scalar shuffle challenge");
-
-    if x.len() != y.len() {
-        return Err(KShuffleError::InvalidR1CSConstruction);
-    }
-    let k = x.len();
-    if k == 1 {
-        cs.add_constraint([(x[0].0, -one), (y[0].0, one)].iter().collect());
-        return Ok(());
-    }
-
-    // Make last x multiplier for i = k-1 and k-2
-    let last_mulx_out = last_multiplier(cs, z, x[k - 1], x[k - 2]);
-
-    // Make multipliers for x from i == [0, k-3]
-    let first_mulx_out = (0..k - 2).rev().fold(last_mulx_out, |prev_out, i| {
-        intermediate_multiplier(cs, z, prev_out?, x[i])
-    })?;
+// Make a gadget that adds constraints to a ConstraintSystem, such that the
+// y variables are constrained to be a valid shuffle of the x variables.
+struct KShuffleGadget {}
 
-    // Make last y multiplier for i = k-1 and k-2
-    let last_muly_out = last_multiplier(cs, z, y[k - 1], y[k - 2]);
+impl KShuffleGadget {
+    fn fill_cs<CS: ConstraintSystem>(cs: &mut CS, x: Vec<Variable>, y: Vec<Variable>) {
```

Comment:
Is there a reason that this needs to take ownership of the `Variable`s, or could it be
```suggestion
    fn shuffle_gadget<CS: ConstraintSystem>(cs: &mut CS, x: &[Variable], y: &[Variable]) {
```
Then the code that calls the gadget in the prove and verify functions can save an alloc by using `split_at` on the `Vec<Variable>` returned from the constraint system

Reply:
Oops, I should have put this comment in the version that's in the test code, not the version that's in the benchmark

Reply:
No reason! That looks good, will update